### ISSUE TITLE
Deadlock on recursive tf.function-decorated function calls

### ISSUE THREAD

**Issue by mayou36** [tensorflow/tensorflow#35540]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 19.10
- TensorFlow installed from (source or binary): binary, conda
- TensorFlow version (use command below): unknown 2.0.0
- Python version: 3.7

**Describe the current behavior**

Recursive calls with a `tf.function` decorated python function results in a deadlock. A minimal example is provided below.

Where it gets stuck: by recursively calling `_maybe_define_function`, which internally requires a lock (line 2118 in `tensorflow_core/python/eager.function.py`). This seems to deadlock when trying to create a graph function again invoking itself.

About use-cases: yes, there are. But yes, in principle, there are "workarounds". But I assume this is not in general intended to produce a deadlock anyway, should it? If so, I would rather propose a loud failure.

**Code to reproduce the issue**
```
@tf.function(autograph=False)
def func1(depth=0):
    if depth > 1:
        return depth
    else:
        return func1(depth + 1)

func1(0)
```

**Comment by gadagashwini** [comment]
Issue is replicating with Tf2.0.
Please find the gist [here](https://colab.sandbox.google.com/gist/gadagashwini/f3c4de7f4c91001fbc81be8c459dcdb8/untitled336.ipynb). Thanks!

**Comment by mayou36** [comment]
While I am only particular familiar with the internals of TF, an `RLock` that allows re-entrance by the same thread may solves the problem.

If the lock is there to avoid _concurrency/parallelism_ to mess with the state of the object, this will work.

If the lock is in place to make sure the state just does one single retrace, this won't.

To unfamiliar with this part of TF to give better ideas.

**Comment by saxenasaurabh** [comment]
Thanks @mayou36 for the detailed description and repro. `tf.function` does not support recursion. Using a `RLock` would lead to stack overflows in the case where the arg is a symbolic `tensor`. I agree we should raise a better error message than deadlocking in this case.

**Comment by mayou36** [comment]
I agree, and I may seem to have hit a rather rare use-case since no one else complained so far.

A possible workaround is to call the `tf.function`-wrapped function (in the recursive scenario) _only_ if it is not yet _inside it's wrapper, otherwise the pure python function is called. This would be a safe method as no deadlocks appear, however not fully using the capability of `tf.function` and somewhat defeating the purpose of recursion as it is just flattened out.

Which is not the golden way and I see that it may does not make sense to implement it, at least not as long as not more people are interested in this an error seems good.

**Comment by bmiselis** [comment]
I can confirm this is the case for me, too.

**Comment by mayou36** [comment]
So two solutions:
- roll out the function (call pure Python function if already inside)
- raise an error if already tracing the function

The second is safer for sure, it somewhat limits the ability to build things; anyway, recursive should not be the way to go but rather `tf.while_loop`

**Comment by mauhcs** [comment]
I had the same issue here.

**Comment by mayou36** [comment]
I would propose to add an error then, `RuntimeError`?

**Comment by ravikyram** [comment]
I am able to reproduce the issue with TF version 2.3-rc1,nightly versions(`2.4.0-dev20200709`) and was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/1455602ef1318f190dda7d466a16162e/untitled106.ipynb).Thanks!

**Comment by fachu000** [comment]
Use case: metalearning using MAML (https://arxiv.org/abs/1703.03400).

Here we need to take a gradient w.r.t. to the initialization of a gradient descent method. This can be implemented naturally in a few lines with a recursive function. 

Unfortunately, due to the limitation expressed in this thread, this works only in Eager mode.

**Comment by sachinprasadhs** [comment]
Was able to reproduce your issue in Tensorflow 2.5 and Tf Nightly version, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/1b23e68f13a8ff47ef0c93361215b34d/35540.ipynb).

**Comment by sanatmpa1** [comment]
Issue still persists in `TF 2.6.0` and in `TF-nightly`. Please take a look at the [gist here](https://colab.research.google.com/gist/sanatmpa1/08f2b91e87240c9de3d337734d485f4d/35540.ipynb)

**Comment by jonas-eschle** [comment]
Great, many thanks, it's really appreciated! That will help to prevent finding deadlocks.

> Note using Tensors as stop condition of recursive tf.function is still not supported.

Sure, that is an inherent feature of the tracing, thanks again
